File: veilid-python/tests/test_routing_context.py
# Routing context veilid tests

import asyncio
import random
import sys
import os

import pytest
import veilid
from veilid.types import OperationId

from .conftest import server_info

##################################################################


@pytest.mark.asyncio
async def test_routing_contexts(api_connection: veilid.VeilidAPI):
    rc = await api_connection.new_routing_context()
    async with rc:
        rcp = await rc.with_privacy(release=False)
        async with rcp:
            rcps = await rcp.with_sequencing(veilid.Sequencing.ENSURE_ORDERED, release=False)
            async with rcps:
                rcpscp = await rcps.with_custom_privacy(veilid.Stability.RELIABLE, release=False)
                await rcpscp.release()


@pytest.mark.asyncio
async def test_routing_context_app_message_loopback():
    # Seriously, mypy?
    app_message_queue: asyncio.Queue = asyncio.Queue()

    async def app_message_queue_update_callback(update: veilid.VeilidUpdate):
        if update.kind == veilid.VeilidUpdateKind.APP_MESSAGE:
            await app_message_queue.put(update)

    hostname, port = server_info()
    api = await veilid.json_api_connect(
        hostname, port, app_message_queue_update_callback
    )
    async with api:
        # purge routes to ensure we start fresh
        await api.debug("purge routes")

        # make a routing context that uses a safety route
        rc = await (await api.new_routing_context()).with_privacy()
        async with rc:

            # make a new local private route
            prl, blob = await api.new_private_route()

            # import it as a remote route as well so we can send to it
            prr = await api.import_remote_private_route(blob)

            # send an app message to our own private route
            message = b"abcd1234"
            await rc.app_message(prr, message)

            # we should get the same message back
            update: veilid.VeilidUpdate = await asyncio.wait_for(
                app_message_queue.get(), timeout=10
            )

            assert isinstance(update.detail, veilid.VeilidAppMessage)
            assert update.detail.message == message


@pytest.mark.asyncio
async def test_routing_context_app_call_loopback():
    app_call_queue: asyncio.Queue = asyncio.Queue()

    async def app_call_queue_update_callback(update: veilid.VeilidUpdate):
        if update.kind == veilid.VeilidUpdateKind.APP_CALL:
            await app_call_queue.put(update)

    hostname, port = server_info()
    api = await veilid.json_api_connect(hostname, port, app_call_queue_update_callback)
    async with api:
        # purge routes to ensure we start fresh
        await api.debug("purge routes")

        # make a routing context that uses a safety route
        rc = await (await api.new_routing_context()).with_privacy()
        async with rc:

            # make a new local private route
            prl, blob = await api.new_private_route()

            # import it as a remote route as well so we can send to it
            prr = await api.import_remote_private_route(blob)

            # send an app message to our own private route
            request = b"abcd1234"
            app_call_task = asyncio.create_task(
                rc.app_call(prr, request), name="app call task"
            )

            # we should get the same request back
            update: veilid.VeilidUpdate = await asyncio.wait_for(
                app_call_queue.get(), timeout=10
            )
            appcall = update.detail

            assert isinstance(appcall, veilid.VeilidAppCall)
            assert appcall.message == request

            # now we reply to the request
            reply = b"qwer5678"
            await api.app_call_reply(appcall.call_id, reply)

            # now we should get the reply from the call
            result = await app_call_task
            assert result == reply


@pytest.mark.asyncio
async def test_routing_context_app_message_loopback_big_packets():

    app_message_queue: asyncio.Queue = asyncio.Queue()

    global got_message
    got_message = 0
    async def app_message_queue_update_callback(update: veilid.VeilidUpdate):
        if update.kind == veilid.VeilidUpdateKind.APP_MESSAGE:
            global got_message
            got_message += 1
            sys.stdout.write("{} ".format(got_message))
            await app_message_queue.put(update)

    sent_messages: set[bytes] = set()

    hostname, port = server_info()
    api = await veilid.json_api_connect(
        hostname, port, app_message_queue_update_callback
    )
    async with api:
        # purge routes to ensure we start fresh
        await api.debug("purge routes")

        # make a routing context that uses a safety route
        rc = await (await (await api.new_routing_context()).with_privacy()).with_sequencing(veilid.Sequencing.ENSURE_ORDERED)
        async with rc:

            # make a new local private route
            prl, blob = await api.new_private_route()

            # import it as a remote route as well so we can send to it
            prr = await api.import_remote_private_route(blob)

            # do this test 1000 times
            for _ in range(1000):

                # send a random sized random app message to our own private route
                message = random.randbytes(random.randint(0, 32768))
                await rc.app_message(prr, message)

                sent_messages.add(message)

            # we should get the same messages back
            print(len(sent_messages))
            for n in range(len(sent_messages)):
                print(n)
                update: veilid.VeilidUpdate = await asyncio.wait_for(
                    app_message_queue.get(), timeout=10
                )
                assert isinstance(update.detail, veilid.VeilidAppMessage)

                assert update.detail.message in sent_messages

@pytest.mark.asyncio
async def test_routing_context_app_call_loopback_big_packets():
    global got_message
    got_message = 0
    
    app_call_queue: asyncio.Queue = asyncio.Queue()

    async def app_call_queue_update_callback(update: veilid.VeilidUpdate):
        if update.kind == veilid.VeilidUpdateKind.APP_CALL:
            await app_call_queue.put(update)

    async def app_call_queue_task_handler(api: veilid.VeilidAPI):
        while True:
            update = await app_call_queue.get()
            
            global got_message
            got_message += 1
            
            sys.stdout.write("{} ".format(got_message))
            sys.stdout.flush()

            await api.app_call_reply(update.detail.call_id, update.detail.message)
        
    hostname, port = server_info()
    api = await veilid.json_api_connect(
        hostname, port, app_call_queue_update_callback
    )
    async with api:
        # purge routes to ensure we start fresh
        await api.debug("purge routes")

        app_call_task = asyncio.create_task(
            app_call_queue_task_handler(api), name="app call task"
        )

        # make a routing context that uses a safety route
        rc = await (await (await api.new_routing_context()).with_privacy()).with_sequencing(veilid.Sequencing.ENSURE_ORDERED)
        async with rc:

            # make a new local private route
            prl, blob = await api.new_private_route()

            # import it as a remote route as well so we can send to it
            prr = await api.import_remote_private_route(blob)

            # do this test 10 times
            for _ in range(10):

                # send a random sized random app message to our own private route
                message = random.randbytes(random.randint(0, 32768))
                out_message = await rc.app_call(prr, message)

                assert message == out_message
        
        app_call_task.cancel()


@pytest.mark.skipif(os.getenv("NOSKIP")!="1", reason="unneeded test, only for performance check")
@pytest.mark.asyncio
async def test_routing_context_app_message_loopback_bandwidth():

    app_message_queue: asyncio.Queue = asyncio.Queue()

    async def app_message_queue_update_callback(update: veilid.VeilidUpdate):
        if update.kind == veilid.VeilidUpdateKind.APP_MESSAGE:
            await app_message_queue.put(True)

    hostname, port = server_info()
    api = await veilid.json_api_connect(
        hostname, port, app_message_queue_update_callback
    )
    async with api:
        # purge routes to ensure we start fresh
        await api.debug("purge routes")

        # make a routing context that uses a safety route
        #rc = await (await (await api.new_routing_context()).with_privacy()).with_sequencing(veilid.Sequencing.ENSURE_ORDERED)
        #rc = await (await api.new_routing_context()).with_privacy()
        rc = await api.new_routing_context()
        async with rc:

            # make a new local private route
            prl, blob = await api.new_private_route()

            # import it as a remote route as well so we can send to it
            prr = await api.import_remote_private_route(blob)

            # do this test 1000 times
            message = random.randbytes(16384)
            for _ in range(10000):

                # send a random sized random app message to our own private route
                await rc.app_message(prr, message)

            # we should get the same number of messages back (not storing all that data)
            for _ in range(10000):
                await asyncio.wait_for(
                    app_message_queue.get(), timeout=10
                )

File: veilid-flutter/example/integration_test/test_routing_context.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';

import 'package:flutter_test/flutter_test.dart';
import 'package:veilid/veilid.dart';

Future<void> testRoutingContexts() async {
  {
    final rc = await Veilid.instance.routingContext();
    rc.close();
  }

  {
    final rc = await Veilid.instance.routingContext();
    final rcp = rc.withDefaultSafety();
    rcp.close();
    rc.close();
  }

  {
    final rc = await Veilid.instance.routingContext();
    final rcp = rc.withSequencing(Sequencing.ensureOrdered);
    rcp.close();
    rc.close();
  }

  {
    final rc = await Veilid.instance.routingContext();
    final rcp = rc.withSafety(const SafetySelectionSafe(
        safetySpec: SafetySpec(
            hopCount: 2,
            stability: Stability.lowLatency,
            sequencing: Sequencing.noPreference)));
    rcp.close();
    rc.close();
  }
  {
    final rc = await Veilid.instance.routingContext();
    final rcp = rc.withSafety(
        const SafetySelectionUnsafe(sequencing: Sequencing.preferOrdered));
    rcp.close();
    rc.close();
  }
}

Future<void> testAppMessageLoopback(Stream<VeilidUpdate> updateStream) async {
  final appMessageQueue = StreamController<VeilidAppMessage>();
  final appMessageSubscription = updateStream.listen((update) {
    if (update is VeilidAppMessage) {
      appMessageQueue.sink.add(update);
    }
  });
  try {
    await Veilid.instance.debug("purge routes");

    // make a routing context that uses a safety route
    final rc = await Veilid.instance
        .safeRoutingContext(sequencing: Sequencing.ensureOrdered);
    try {
      // make a new local private route
      final prl = await Veilid.instance.newPrivateRoute();
      try {
        // import it as a remote route as well so we can send to it
        final prr = await Veilid.instance.importRemotePrivateRoute(prl.blob);
        try {
          // send an app message to our own private route
          final message = utf8.encode("abcd1234");
          await rc.appMessage(prr, message);

          // we should get the same message back
          final update = await appMessageQueue.stream.first;
          expect(update.message, equals(message));
          expect(update.routeId, isNotNull);
        } finally {
          await Veilid.instance.releasePrivateRoute(prr);
        }
      } finally {
        await Veilid.instance.releasePrivateRoute(prl.routeId);
      }
    } finally {
      rc.close();
    }
  } finally {
    await appMessageSubscription.cancel();
  }
}

Future<void> testAppCallLoopback(Stream<VeilidUpdate> updateStream) async {
  final appCallQueue = StreamController<VeilidAppCall>();
  final appMessageSubscription = updateStream.listen((update) {
    if (update is VeilidAppCall) {
      appCallQueue.sink.add(update);
    }
  });
  try {
    await Veilid.instance.debug("purge routes");

    // make a routing context that uses a safety route
    final rc = await Veilid.instance
        .safeRoutingContext(sequencing: Sequencing.ensureOrdered);
    try {
      // make a new local private route
      final prl = await Veilid.instance.newPrivateRoute();
      try {
        // import it as a remote route as well so we can send to it
        final prr = await Veilid.instance.importRemotePrivateRoute(prl.blob);
        try {
          // send an app call to our own private route
          final message = utf8.encode("abcd1234");
          final appCallFuture = rc.appCall(prr, message);

          // we should get the same call back
          final update = await appCallQueue.stream.first;
          final appcallid = update.callId;

          expect(update.message, equals(message));
          expect(update.routeId, isNotNull);

          // now we reply to the request
          final reply = utf8.encode("qwer5678");
          await Veilid.instance.appCallReply(appcallid, reply);

          // now we should get the reply from the call
          final result = await appCallFuture;
          expect(result, equals(reply));
        } finally {
          await Veilid.instance.releasePrivateRoute(prr);
        }
      } finally {
        await Veilid.instance.releasePrivateRoute(prl.routeId);
      }
    } finally {
      rc.close();
    }
  } finally {
    await appMessageSubscription.cancel();
  }
}

Future<void> testAppMessageLoopbackBigPackets(
    Stream<VeilidUpdate> updateStream) async {
  final appMessageQueue = StreamController<VeilidAppMessage>();
  final appMessageSubscription = updateStream.listen((update) {
    if (update is VeilidAppMessage) {
      appMessageQueue.sink.add(update);
    }
  });

  final sentMessages = <String>{};
  final random = Random.secure();
  final cs = await Veilid.instance.bestCryptoSystem();

  try {
    await Veilid.instance.debug("purge routes");

    // make a routing context that uses a safety route
    final rc = await Veilid.instance
        .safeRoutingContext(sequencing: Sequencing.ensureOrdered);
    try {
      // make a new local private route
      final prl = await Veilid.instance.newPrivateRoute();
      try {
        // import it as a remote route as well so we can send to it
        final prr = await Veilid.instance.importRemotePrivateRoute(prl.blob);
        try {
          for (var i = 0; i < 5; i++) {
            // send an app message to our own private route
            final message = await cs.randomBytes(random.nextInt(32768));
            await rc.appMessage(prr, message);
            sentMessages.add(base64Url.encode(message));
          }

          final appMessageQueueIterator =
              StreamIterator(appMessageQueue.stream);

          // we should get the same messages back
          for (var i = 0; i < sentMessages.length; i++) {
            if (await appMessageQueueIterator.moveNext()) {
              final update = appMessageQueueIterator.current;
              expect(sentMessages.contains(base64Url.encode(update.message)),
                  isTrue);
            } else {
              fail("not enough messages in the queue");
            }
          }
        } finally {
          await Veilid.instance.releasePrivateRoute(prr);
        }
      } finally {
        await Veilid.instance.releasePrivateRoute(prl.routeId);
      }
    } finally {
      rc.close();
    }
  } finally {
    await appMessageSubscription.cancel();
  }
}

Future<void> testAppCallLoopbackBigPackets(
    Stream<VeilidUpdate> updateStream) async {
  final appCallQueue = StreamController<VeilidAppCall>();
  final appMessageSubscription = updateStream.listen((update) {
    if (update is VeilidAppCall) {
      appCallQueue.sink.add(update);
    }
  });
  final appCallQueueHandler = () async {
    await for (final update in appCallQueue.stream) {
      await Veilid.instance.appCallReply(update.callId, update.message);
    }
  }();

  final sentMessages = <String>{};
  final random = Random.secure();
  final cs = await Veilid.instance.bestCryptoSystem();

  try {
    await Veilid.instance.debug("purge routes");

    // make a routing context that uses a safety route
    final rc = (await Veilid.instance.routingContext())
        .withSequencing(Sequencing.ensureOrdered, closeSelf: true);
    try {
      // make a new local private route
      final prl = await Veilid.instance
          .newCustomPrivateRoute(Stability.reliable, Sequencing.ensureOrdered);
      try {
        // import it as a remote route as well so we can send to it
        final prr = await Veilid.instance.importRemotePrivateRoute(prl.blob);
        try {
          for (var i = 0; i < 5; i++) {
            // send an app message to our own private route
            final message = await cs.randomBytes(random.nextInt(32768));
            final outmessage = await rc.appCall(prr, message);
            expect(message, equals(outmessage));
          }

          final appMessageQueueIterator = StreamIterator(appCallQueue.stream);

          // we should get the same messages back
          for (var i = 0; i < sentMessages.length; i++) {
            if (await appMessageQueueIterator.moveNext()) {
              final update = appMessageQueueIterator.current;
              expect(sentMessages.contains(base64Url.encode(update.message)),
                  isTrue);
            } else {
              fail("not enough messages in the queue");
            }
          }
        } finally {
          await Veilid.instance.releasePrivateRoute(prr);
        }
      } finally {
        await Veilid.instance.releasePrivateRoute(prl.routeId);
      }
    } finally {
      rc.close();
    }
  } finally {
    await appMessageSubscription.cancel();
  }
  await appCallQueue.close();
  await appCallQueueHandler;
}
